src/shortcuts/enter.js

```diff
@@ -0,0 +1,63 @@
+import React from 'react'
+import { isMobile } from '../browser'
+
+// util
+import {
+  getOffsetWithinContent,
+  headValue,
+  isDocumentEditable,
+} from '../util'
+
+import { newThought } from '../action-creators'
+
+// eslint-disable-next-line jsdoc/require-jsdoc
+const Icon = ({ fill = 'black', size = 20, style }) => <svg version="1.1" className="icon" xmlns="http://www.w3.org/2000/svg" width={size} height={size} fill={fill} style={style} viewBox="0 0 19.481 19.481" enableBackground="new 0 0 19.481 19.481">
+  <g>
+    <path d="m10.201,.758l2.478,5.865 6.344,.545c0.44,0.038 0.619,0.587 0.285,0.876l-4.812,4.169 1.442,6.202c0.1,0.431-0.367,0.77-0.745,0.541l-5.452-3.288-5.452,3.288c-0.379,0.228-0.845-0.111-0.745-0.541l1.442-6.202-4.813-4.17c-0.334-0.289-0.156-0.838 0.285-0.876l6.344-.545 2.478-5.864c0.172-0.408 0.749-0.408 0.921,0z" />
+  </g>
+</svg>
+
+/** The newThought command handler that does some pre-processing before handing off to newThought. */
+const exec = (dispatch, getState, e, { type }) => {
+  const state = getState()
+  const { cursor } = state
+
+  // outdent instead of creating new thought if current cursor is an empty thought
+  if (cursor && cursor.length > 2 && headValue(cursor).length === 0) {
+    dispatch({ type: 'outdent' })
+  }
+  else {
+    // Note: Jest triggers new thought with windowEvent which has window as target causing getOffsetWithinContent to fail
+    const isTargetHTMLElement = e.target instanceof HTMLElement
+
+    // Note: e.target should be a HTMLElement and a content editable node
+    const offset = isTargetHTMLElement ? getOffsetWithinContent(e.target) : 0
+
+    // prevent split on gesture
+    dispatch(newThought({ offset, preventSplit: type === 'gesture' }))
+  }
+}
+
+export default {
+  id: 'enter',
+  name: 'Enter',
+  description: 'Create a new thought or outdent if focused thought is empty.',
+  keyboard: { key: 'Enter' },
+  gesture: 'rd',
+  svg: Icon,
+  canExecute: () => isDocumentEditable(),
+  exec
+}
+
+// add aliases to help with mis-swipes since MultiGesture does not support diagonal swipes
+export const enterAliases = {
+  id: 'enterAliases',
+  name: 'Enter',
+  hideFromInstructions: true,
+  gesture: ['rdld', 'rdldl', 'rdldld', 'rldl', 'rldld', 'rldldl'],
+  // on mobile, the shift key should cause a normal newThought, not newThoughtAbove
+  // smuggle it in with the aliases
+  ...isMobile ? { keyboard: { key: 'Enter', shift: true } } : null,
+  canExecute: () => isDocumentEditable(),
+  exec
+}
```

Comment:
I'd invite you to think more about the semantics of this. We want to think about the architectural semantics at all levels, just like the reasoning we took to separate `newThought` and `enter`.

**Why would gestures be bound to a shortcut called `enter`?**

This catches one's attention because it is semantically incongruous. Once investigated it also reveals an unintended effect: Doing the gesture for `newThought` on an empty thought will cause an outdent, though the intention is that this will only happen when the Enter key is pressed.

I'd like to hear a proposal to resolve the incongruity and prevent the outdent on gesture.

Reply:
> Why would gestures be bound to a shortcut called `enter`?

We shouldn't name it based on `keyboard` or `gesture`. We should go with something like `newThoughtOrOutdent` like you mentioned earlier.

Reply:
> Doing the gesture for `newThought` on an empty thought will cause an outdent, though the intention is that this will only happen when the Enter key is pressed.

We should encapsulate individual actions into their own separate `exec` functions. Also we should create separate functions to check if the certain actions can be executed. This way we can write main `exec` function that can decide which action to trigger based on certain conditions or `type` of event (i.e `keyboard` or `gesture`). Even if the aliases behave differently we can create a exec function that only runs certain actions with very little code.

```js
{
  exec : (dispatch, getState, e, { type }) => {
    if (shouldExecOutdent(getState) && type === 'keyboard') dispatch({ type: 'outdent' })
    else newThoughtExec(dispatch, getState, e, { type })
  }
}

// for alias that doesn't need to trigger outdent
{
  exec: newThoughtExec
}
```